File: 02/js/main.js
const fs = require('fs')
const {join} = require('path')

const file = fs.readFileSync(join(__dirname, '../input.txt'), 'utf8')
const inputs = file.split('\n').slice(0, -1)


// Problem 1
{
  let forward = 0
  let down = 0

  for (let i = 0; i < inputs.length; i++) {
    const [dir, str] = inputs[i].split(' ')
    const num = Number(str)

    switch (dir) {
      case 'up': down -= num; break;
      case 'down': down += num; break;
      case 'forward': forward += num; break;

      default: console.log(`Unknown dir: ${dir} at: ${i}`);
    }
  }

  console.log(`Count 1: ${forward * down}`);
}

// Problem 2
{
  let forward = 0
  let down = 0
  let aim = 0

  for (let i = 0; i < inputs.length; i++) {
    const [dir, str] = inputs[i].split(' ')
    const num = Number(str)

    switch (dir) {
      case 'up': aim -= num; break;
      case 'down': aim += num; break;
      case 'forward': forward += num; down += aim * num; break;

      default: console.log(`Unknown dir: ${dir} at: ${i}`);
    }
  }

  console.log(`Count 2: ${forward * down}`);
}

File: 02/c/main.c
#include <stdio.h>
#include <assert.h>
#include <string.h>

#define N 2048

struct inst {
  int num;
  char dir[10];
};

struct inst xs[N];
int xs_size = 0;

int main(int argc, char *argv[]) {
  assert(argc == 2 && "Must provide one argument => input.txt path");

  printf("Input file: %s\n", argv[1]);
  FILE *file = fopen(argv[1], "r");

  int number = 0;
  char str[10] = "";

  while (fscanf(file, "%s", str) == 1 && fscanf(file, "%d", &number) == 1) {
    xs[xs_size].num = number;
    strcpy(xs[xs_size++].dir, str);
  }


  // Problema 1
  {
    int down = 0;
    int forward = 0;

    for (int i = 0; i < xs_size; i++) {
      if (strcmp(xs[i].dir, "up") == 0) {
        down -= xs[i].num;
      } else if (strcmp(xs[i].dir, "down") == 0) {
        down += xs[i].num;
      } else if (strcmp(xs[i].dir, "forward") == 0) {
        forward += xs[i].num;
      } else {
        printf("Unknown dir: %s at: %d", xs[i].dir, i);
      }
    }

    printf("\nCount 1: %d", down * forward);
  }

  // Problema 2
  {
    int down = 0;
    int forward = 0;
    int aim = 0;

    for (int i = 0; i < xs_size; i++) {
      if (strcmp(xs[i].dir, "up") == 0) {
        aim -= xs[i].num;
      } else if (strcmp(xs[i].dir, "down") == 0) {
        aim += xs[i].num;
      } else if (strcmp(xs[i].dir, "forward") == 0) {
        forward += xs[i].num;
        down += aim * xs[i].num;
      } else {
        printf("Unknown dir: %s at: %d", xs[i].dir, i);
      }
    }

    printf("\nCount 1: %d", down * forward);
  }

  fclose(file);

  printf("\n");
  return 0;
}
